[SPARK-53803][ML][Feature] Added ArimaRegression for time series forecasting in MLlib

anandexplore opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'common/utils/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request? 

This pull request adds a new feature called ArimaRegression to Spark MLlib under org.apache.spark.ml.regression.
It brings the ARIMA (AutoRegressive Integrated Moving Average) model for one-variable (univariate) time series forecasting, along with a matching model class ArimaRegressionModel.

The update includes:
Scala code for ArimaRegression and ArimaRegressionModel
Support for ARIMA parameters: p, d, and q
PySpark API bindings for both classes
Unit tests in Scala and Python
Model save/load support using MLWritable and MLReadable
Example usage in examples/ml/ArimaRegressionExample.scala

### Why are the changes needed?
Currently, Spark MLlib does not have built-in tools for time-series forecasting.
ARIMA is one of the most common models for predicting trends in time-based data.
Adding this feature allows Spark users to perform forecasting directly within MLlib, without needing outside Python libraries. It also makes Spark’s machine learning toolkit more complete.

### Does this PR introduce _any_ user-facing change?
Yes.
New APIs are available in both Scala and Python:
org.apache.spark.ml.regression.ArimaRegression
org.apache.spark.ml.regression.ArimaRegressionModel
pyspark.ml.regression.ArimaRegression
pyspark.ml.regression.ArimaRegressionModel

These follow standard Spark ML APIs and work with Pipelines, ParamMaps, save/load, and transform().
### How was this patch tested?
Tests were added in:

Scala (ArimaRegressionSuite.scala) for:
Model fitting and transforming
Parameter defaults and setters
Save/load functions
Python (test_regression.py) for PySpark interface
Manual testing was also done in both:
spark-shell (Scala)
pyspark (Python)

Manual Tested
Scala:
import org.apache.spark.ml.regression.ArimaRegression
import org.apache.spark.sql.SparkSession

val spark = SparkSession.builder().appName("ArimaRegressionExample").getOrCreate()
import spark.implicits._
val data = Seq(100.0, 102.5, 101.0, 104.0, 107.5, 110.0).toDF("value")
val arima = new ArimaRegression()
  .setP(1)
  .setD(1)
  .setQ(1)
val model = arima.fit(data)
val forecast = model.transform(data)
forecast.show(false)

Python:
from pyspark.ml.regression import ArimaRegression
from pyspark.sql import SparkSession
spark = SparkSession.builder.appName("ArimaRegressionExample").getOrCreate()
data = [(100.0,), (102.5,), (101.0,), (104.0,), (107.5,), (110.0,)]
df = spark.createDataFrame(data, ["value"])
arima = ArimaRegression(p=1, d=1, q=1)
model = arima.fit(df)
forecast = model.transform(df)
forecast.show(truncate=False)
Predictions and output schema were checked for correctness

### Was this patch authored or co-authored using generative AI tooling?
No.

anandexplore committed:
[SPARK-53803][ML][Feature] Add ArimaRegression for time series forecasting in MLlib

Add ArimaRegression for time series forecasting in MLlib

anandexplore committed:
Merge branch 'apache:master' into master

anandexplore commented:
@sryza  Could you review this commit

anandexplore committed:
Merge branch 'apache:master' into master

anandexplore committed:
Merge branch 'apache:master' into master

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/ArimaRegression.scala:
```diff
@@ -0,0 +1,58 @@
+
+  override def fit(dataset: Dataset[_]): ArimaRegressionModel = {
+    // Dummy: assumes data is ordered with one feature column "y"
+    val ts = dataset.select("y").rdd.map(_.getDouble(0)).collect()
```
If this implementation needs to collect all the dataset, then I  doesn't think it match the requirement

> Be highly scalable

from the `MLlib-specific contribution guidelines` [here](https://spark.apache.org/contributing.html)

zhengruifeng replied:
And I think a more mature usage is using https://www.statsmodels.org/stable/generated/statsmodels.tsa.arima.model.ARIMA.html with pyspark

zhengruifeng replied:
> I’ll refactor the logic to keep computations distributed

@anandexplore before working on it, let's wait for more input from @WeichenXu123 @srowen @mengxr on the need of new algorithm

anandexplore replied:
@WeichenXu123 @srowen Please review the PR and ’ll wait for your additional input to add new model..

anandexplore replied:
@zhengruifeng Just checking in — any update on the refactor to keep the computations distributed? I’ve also pushed a few updates on my side, so please take a look when you get a chance. The PR has been pending for a while.

cc:@WeichenXu123 @srowen

zhengruifeng replied:
@anandexplore TBH, I am not sure whether we should add this in mllib.
I suspect users should just use existing python/R libs for time series.

And even if we decide to implement it in mllib, I believe we need a scalable one writen in Scala. (not by a Python UDF https://github.com/apache/spark/pull/52519/files#diff-60237792895c7800896b19fc382faa926089d3b88ba96d6bbe03353aec12a73fR58-R75)

anandexplore replied:
[zhengruifeng] Sure will hold on this untill get feedback from ML committers.

anandexplore committed:
Merge branch 'apache:master' into master

anandexplore committed:
[SPARK-53803][ML][Feature] Add ArimaRegression for time series forecasting

[SPARK-53803][ML][Feature] Add ArimaRegression for time series forecasting

anandexplore committed:
[SPARK-53803][ML][Feature] Add ArimaRegression for time series forecasting

[SPARK-53803][ML][Feature] Add ArimaRegression for time series forecasting

anandexplore committed:
[SPARK-53803][ML][Feature] Added ArimaRegression - Added ScalaType fix

[SPARK-53803][ML][Feature] Added ArimaRegression - Added ScalaType fix

anandexplore committed:
Open [SPARK-53803][ML][Feature] Added ArimaRegression for time series forecasting

Open [SPARK-53803][ML][Feature] Added ArimaRegression for time series forecasting

anandexplore committed:
[SPARK-53803][ML][Feature] Add ArimaRegression for time series forecasting
